server/controllers/api_controller.go

```diff
@@ -0,0 +1,233 @@
+	ProjectApplyCommandRunner events.ProjectApplyCommandRunner
+	RepoAllowlistChecker      *events.RepoAllowlistChecker
+	Scope                     tally.Scope
+	VCSClient                 vcs.Client
+}
+
+type APIRequest struct {
+	Repository string `validate:"required"`
+	Ref        string `validate:"required"`
+	Type       string `validate:"required"`
+	Projects   []string
+	Paths      []struct {
+		Directory string
+		Workspace string
+	}
+}
+
+func (a *APIRequest) getCommands(ctx *command.Context, cmdBuilder func(*command.Context, *events.CommentCommand) ([]command.ProjectContext, error)) ([]command.ProjectContext, error) {
+	cc := make([]*events.CommentCommand, 0)
+
+	for _, project := range a.Projects {
+		cc = append(cc, &events.CommentCommand{
+			ProjectName: project,
+		})
+	}
+	for _, path := range a.Paths {
+		cc = append(cc, &events.CommentCommand{
+			RepoRelDir: strings.TrimRight(path.Directory, "/"),
+			Workspace:  path.Workspace,
+		})
+	}
+
+	cmds := make([]command.ProjectContext, 0)
+	for _, commentCommand := range cc {
+		projectCmds, err := cmdBuilder(ctx, commentCommand)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build command: %v", err)
+		}
+		cmds = append(cmds, projectCmds...)
+	}
+
+	return cmds, nil
+}
+
+func (a *APIController) apiReportError(w http.ResponseWriter, code int, err error) {
+	response, _ := json.Marshal(map[string]string{
+		"error": err.Error(),
+	})
+	a.respond(w, logging.Warn, code, string(response))
+}
+
+func (a *APIController) Plan(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	request, ctx, code, err := a.apiParseAndValidate(r)
+	if err != nil {
+		a.apiReportError(w, code, err)
+		return
+	}
+
+	result, err := a.apiPlan(request, ctx)
+	if err != nil {
+		a.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+	defer a.Locker.UnlockByPull(ctx.HeadRepo.FullName, 0) // nolint: errcheck
+	if result.HasErrors() {
+		code = http.StatusInternalServerError
+	}
+
+	// TODO: make a better response
+	response, err := json.Marshal(result)
+	if err != nil {
+		a.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+	a.respond(w, logging.Debug, code, string(response))
+}
+
+func (a *APIController) Apply(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	request, ctx, code, err := a.apiParseAndValidate(r)
+	if err != nil {
+		a.apiReportError(w, code, err)
+		return
+	}
+
+	// We must first make the plan for all projects
+	_, err = a.apiPlan(request, ctx)
+	if err != nil {
+		a.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+	defer a.Locker.UnlockByPull(ctx.HeadRepo.FullName, 0) // nolint: errcheck
+
+	// We can now prepare and run the apply step
+	result, err := a.apiApply(request, ctx)
+	if err != nil {
+		a.apiReportError(w, http.StatusInternalServerError, err)
+		return
+	}
+	if result.HasErrors() {
+		code = http.StatusInternalServerError
+	}
```

Comment:
Seems like the apiApply function call and the next 2 if statements are the only differences when compared to the Plan function. Can we combine these 2 functions together and use an if statement to check if it should do an apply or not? That way it would DRY up the code a bit

Reply:
I understand there is plenty of room for improvement, but I'm less sure about the specific change you are proposing. Could you suggest the change you described to avoid confusion? Much appreciated!